tests.js

```diff
@@ -704,6 +690,297 @@ describe('Tarn', () => {
+      validate(resource) {
+        return Promise.delay(10).then(() => true);
+      },
+      destroy(resource) {},
+      min: 0,
+      max: 2,
+      acquireTimeoutMillis: 150
+    });
+
+    return pool.acquire().promise.then(res => {
+      pool.release(res);
+    });
+  });
+
+  it('should destroy resource if validate times out', () => {
+    let createCalled = 0;
+    let destroyCalled = 0;
+    let destroyed = null;
+    let validated = null;
+
+    pool = new Pool({
+      create(callback) {
+        callback(null, { a: createCalled++, n: 0 });
+      },
+      validate(resource) {
+        validated = resource;
+        return Promise.delay(10000).then(() => true);
+      },
+      destroy(resource) {
+        destroyed = resource;
+        ++destroyCalled;
+      },
+      min: 0,
+      max: 2,
+      acquireTimeoutMillis: 10
+    });
+
+    // NOTE: it is completely ok that acquire rejects before destroy for
+    // the related resource has been called. Thats why we need to wait
+    // a moment before checking the expected results.
+    return pool
+      .acquire()
+      .promise.catch(err => {
+        return Promise.delay(1);
+      })
+      .then(() => {
+        expect(destroyCalled).to.be(1);
+        expect(createCalled).to.be(1);
+        expect(destroyed).to.be(validated);
+      });
+  });
+
+  it('should allow to abort pending acquire even if validation is not ready', () => {
+    let createCalled = 0;
+    let destroyCalled = 0;
+    let destroyed = null;
+    let validated = null;
+
+    pool = new Pool({
+      create(callback) {
+        callback(null, { a: createCalled++, n: 0 });
+      },
+      validate(resource) {
+        validated = resource;
+        return Promise.delay(10000).then(() => true);
+      },
+      destroy(resource) {
+        destroyed = resource;
+        ++destroyCalled;
+      },
+      min: 0,
+      max: 2,
+      acquireTimeoutMillis: 15000
+    });
+
+    let pending = pool.acquire();
+
+    return Promise.delay(20)
+      .then(() => {
+        pending.abort();
+        return pending.promise;
+      })
+      .catch(err => {})
+      .then(() => {
+        // TODO: validation should immediately return false?
+        expect(destroyCalled).to.be(1);
+        expect(createCalled).to.be(1);
+        expect(destroyed).to.be(validated);
+      });
+  });
+
+  it('should destroy resource if validate throws synchronus exception', () => {
+    let createCalled = 0;
+    let destroyCalled = 0;
+    let numberOfFailingValidates = 5;
+
+    pool = new Pool({
+      create(callback) {
+        callback(null, { a: createCalled++, n: 0 });
+      },
+      validate(resource) {
+        if (numberOfFailingValidates > 0) {
+          numberOfFailingValidates--;
+          throw new Error('fail');
+        }
+
+        return true;
+      },
+      destroy(resource) {
+        ++destroyCalled;
+      },
+      min: 0,
+      max: 2,
+      acquireTimeoutMillis: 100
+    });
+
+    return pool
```

Comment:
Why not async/await? We are at Node 8 already for Tarn

Reply:
I didn't want to mix async / await in yet in this PR.